Nodejs/Product/Nodejs/TypeScriptHelpers/TypeScriptCompile.cs

```diff
@@ -0,0 +1,107 @@
+        {
+            if (!TsConfigJsonFactory.IsTsConfigJsonFile(filePath) || !Path.IsPathRooted(filePath))
+            {
+                throw new ArgumentException($"{nameof(filePath)} has to be a rooted 'tsconfig.json' file.");
+            }
+
+            return await CompileAsync($"-p \"{filePath}\"", workingDir: Path.GetDirectoryName(filePath), redirector: new CompileRedirector(outputPane));
+        }
+
+        private static async Task<bool> CompileAsync(string arguments, string workingDir, Redirector redirector)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(arguments), $"{nameof(arguments)} should not be empty.");
+
+            var pathToTsc = Path.Combine(TypeScriptCompilerLocator.GetDefaultVersion(), CompilerExe);
+
+            redirector?.WriteLine($"=== {Resources.TscBuildStarted}: {pathToTsc} {arguments} ===");
+
+            using (var process = ProcessOutput.Run(
+                pathToTsc,
+                new[] { arguments },
+                workingDir,
+                env: null,
+                visible: false,
+                redirector: redirector,
+                quoteArgs: false,
+                outputEncoding: Encoding.UTF8))
+            {
+                var whnd = process.WaitHandle;
+                if (whnd == null)
+                {
+                    // Process failed to start, and any exception message has
+                    // already been sent through the redirector
+                    redirector.WriteErrorLine(string.Format(Resources.TscBuildError, pathToTsc));
+                    return false;
+                }
+                else
+                {
+                    var finished = await Task.Run(() => whnd.WaitOne());
+                    if (finished)
+                    {
+                        Debug.Assert(process.ExitCode.HasValue, "tsc.exe process has not really exited");
+                        // there seems to be a case when we're signalled as completed, but the
+                        // process hasn't actually exited
+                        process.Wait();
+
+                        redirector.WriteErrorLine($"==== {Resources.TscBuildCompleted} ====");
+
+                        return process.ExitCode == 0;
+                    }
+                    else
+                    {
+                        process.Kill();
+                        redirector.WriteErrorLine($"==== {Resources.TscBuildCancelled} ====");
+
+                        return false;
+                    }
+                }
+            }
+        }
+
+        private sealed class CompileRedirector : Redirector
+        {
+            private readonly OutputPaneWrapper outputPane;
+
+            public CompileRedirector(OutputPaneWrapper outputPane)
+            {
+                this.outputPane = outputPane;
+            }
+
+            public override void WriteErrorLine(string line)
+            {
+                this.outputPane.WriteLine(line, OutputWindowTarget.Tsc);
```

Comment:
>OutputWindowTarget.Tsc [](start = 48, length = 22)

Is this were a parameter to the ctor, the class could be reused for the other window targets.